### PR TITLE
EDSC-3861: Implementing map legend

### PR DESCRIPTION
**Pull request by trevorlang**
# Overview

### What is the feature?

Adds the legend/colormap to the map

### What is the Solution?

Use the GIBS tag to find layers, request the colormaps for those layers, and render them in the Legend.

### What areas of the application does this impact?

Map

# Testing

### Reproduction steps

- **Environment for testing:** PROD
- **Collection to test with:** C1996881146-POCLOUD

1. Load C1996881146-POCLOUD
- [ ] Verify colormap shows in the geographic projection
- [ ] Verify the colormap does not show on arctic/polar projections
- [ ] Verity the map imagery loads as expected
2. Spot check various map imagery

### Attachments

image

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: cypress/e2e/paths/search/search.cy.js [hunk 1]
```diff
@@ -259,7 +259,8 @@ describe('Path /search', () => {
         getByTestId('spatial-display_point').should('have.value', '4.33676,65.44171')
 
         // Test leaflet has drawn the shape correctly
-        cy.get('.leaflet-marker-pane img').should('have.attr', 'style', 'margin-left: -12px; margin-top: -41px; width: 25px; height: 41px; transform: translate3d(1165px, 402px, 0px); z-index: 402;')
+        // cy.get('.leaflet-marker-pane img').should('have.attr', 'style', 'margin-left: -12px; margin-top: -41px; width: 25px; height: 41px; transform: translate3d(1165px, 402px, 0px); z-index: 402;')
```
Comment (trevorlang) [inline cypress/e2e/paths/search/search.cy.js:262]: Remove commented code

File: serverless/src/generateColorMaps/getProjectionCapabilities.js [hunk 2]
```diff
@@ -58,7 +56,9 @@ export const getProjectionCapabilities = async (projection) => {
         .where({ product: layer['ows:Identifier'] })
 
       const metadataLinks = layer['ows:Metadata'] || []
+      console.log('metadataLinks', metadataLinks)
```
Comment (trevorlang) [inline serverless/src/generateColorMaps/getProjectionCapabilities.js:59]: Remove

File: static/src/js/components/Legend/Legend.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
+import React, { useEffect, useRef, useState } from 'react'
+import PropTypes from 'prop-types'
+import hexToRgba from 'hex-to-rgba'
+
+import './Legend.scss'
+import classNames from 'classnames'
+
```
Comment (trevorlang) [inline static/src/js/components/Legend/Legend.js:7]: Comment this component

File: static/src/js/components/Legend/Legend.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        ref={barRef}
+        className="legend__bar"
+        onMouseMove={(e) => onMouseMove(e)}
+        onMouseLeave={(e) => onMouseLeave(e)}
+      />
+      <div className="legend__meta">
+        {
+          isFocused
+            ? (
+              <div className="legend__focus">
```
Comment (trevorlang) [inline static/src/js/components/Legend/Legend.js:99]: Do this safer

File: static/src/js/components/Legend/__tests__/Legend.test.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import React from 'react'
+import '@testing-library/jest-dom'
+import {
+  render,
+  screen
+  // waitFor,
```
Comment (trevorlang) [inline static/src/js/components/Legend/__tests__/Legend.test.js:6]: Remove commented code

File: static/src/js/containers/MapContainer/MapContainer.js [hunk 6]
```diff
@@ -173,7 +181,7 @@ export const MapContainer = (props) => {
     })
   }
 
-  const handleProjectionSwitching = (projection) => {
+  const handleProjectionSwitching = useCallback((projection) => {
```
Comment (trevorlang) [inline static/src/js/containers/MapContainer/MapContainer.js:184]: Double check if we need to do this now

File: static/src/js/containers/MapContainer/MapWrapper.js [hunk 7]
```diff
@@ -78,6 +89,56 @@ const MapWrapper = (props) => {
...
   }, [])
 
+  const [gibsLayer, setGibsLayer] = useState('')
+  const [colorMap, setColorMap] = useState({})
+
+  useEffect(() => {
+    const { tags } = focusedCollectionMetadata
+    const gibsTag = getValueForTag('gibs', tags)
+
+    setGibsLayer({
```
Comment (trevorlang) [inline static/src/js/containers/MapContainer/MapWrapper.js:99]: Remove some debugging code

File: static/src/js/containers/MapContainer/MapWrapper.js [hunk 8]
```diff
@@ -78,6 +89,56 @@ const MapWrapper = (props) => {
...
+  useEffect(() => {
+    const getColorMap = async () => {
+      if (
+        gibsLayer
+        && gibsLayer.product
+        && hasGibsLayerForProjection(gibsLayer, projection)
+        && !colorMap[projection]
+      ) {
+        try {
+          const response = await fetch(`http://localhost:3001/dev/colormaps/${gibsLayer.product}`)
```
Comment (trevorlang) [inline static/src/js/containers/MapContainer/MapWrapper.js:126]: Chance to environment variables

File: static/src/js/containers/MapContainer/MapWrapper.js [hunk 9]
```diff
@@ -78,6 +92,50 @@ const MapWrapper = (props) => {
...
+  useEffect(() => {
+    const getColorMap = async () => {
+      if (
+        gibsLayer
+        && gibsLayer.product
+        && hasGibsLayerForProjection(gibsLayer, projection)
+        && !colorMap[projection]
+      ) {
+        try {
+          const response = await fetch(`${apiHost}/colormaps/${gibsLayer.product}`)
```
Comment (macrouch) [inline static/src/js/containers/MapContainer/MapWrapper.js:123]: Should this be an action that puts the values into Redux?

**Review by daniel-zamora — APPROVED** [review]

**Review by eudoroolivares2016 — APPROVED** [review]

**Pull request merged**